File: src/lib/rename/service.rs
use super::model::{RenameError, RenameOperation};
use super::ports::{RenameService, TemplateContext};
use log::{debug, info};
use std::fs;
use std::path::PathBuf;
use tera::Tera;

pub struct FileRenameService {
    base_dir: PathBuf,
}

impl FileRenameService {
    pub fn new(base_dir: PathBuf) -> Self {
        Self { base_dir }
    }
}

impl RenameService for FileRenameService {
    fn find_files(&self, mask: &str) -> Result<Vec<PathBuf>, RenameError> {
        info!("searching for files with mask: {}", mask);

        let mask_lower = mask.to_lowercase();
        let mut found_files = Vec::new();

        let entries = fs::read_dir(&self.base_dir).map_err(|e| {
            RenameError::IoError(std::io::Error::new(
                e.kind(),
                format!("failed to read directory: {}", e),
            ))
        })?;

        for entry in entries {
            let entry = entry?;
            let path = entry.path();

            if !path.is_file() {
                continue;
            }

            if let Some(ext) = path.extension() {
                if ext.to_str() != Some("srt") {
                    continue;
                }
            } else {
                continue;
            }

            if let Some(filename) = path.file_name().and_then(|n| n.to_str())
                && filename.to_lowercase().contains(&mask_lower)
            {
                debug!("found file: {:?}", path);
                found_files.push(path);
            }
        }

        if found_files.is_empty() {
            return Err(RenameError::NoFilesFound(mask.to_string()));
        }

        // Sort alphabetically
        found_files.sort();

        info!("found {} files", found_files.len());
        Ok(found_files)
    }

    fn prepare_rename_operations(
        &self,
        files: Vec<PathBuf>,
        template: &str,
        context: &TemplateContext,
        series_mode: bool,
    ) -> Result<Vec<RenameOperation>, RenameError> {
        info!("preparing rename operations");
        debug!("series_mode: {}", series_mode);
        debug!("template: {}", template);

        let mut tera = Tera::default();
        tera.add_raw_template("filename", template)
            .map_err(|e| RenameError::InvalidTemplate(e.to_string()))?;

        let mut operations = Vec::new();

        for (index, file_path) in files.iter().enumerate() {
            let mut tera_context = tera::Context::new();

            if let Some(name) = &context.name {
                tera_context.insert("name", name);
            }

            if let Some(season) = context.season {
                tera_context.insert("season", &format!("{:02}", season));
            }

            if let Some(language) = &context.language {
                tera_context.insert("language", language);
            }

            if let Some(separator) = &context.separator {
                tera_context.insert("separator", separator);
            }

            if series_mode {
                let episode_number = index + 1;
                tera_context.insert("episode", &format!("{:02}", episode_number));
                debug!("series mode: file {} -> episode {}", index, episode_number);
            }

            let new_name = tera
                .render("filename", &tera_context)
                .map_err(|e| RenameError::TemplateError(e.to_string()))?;

            debug!("new name for {:?}: {}", file_path, new_name);

            let new_path = self.base_dir.join(&new_name);
            let collision = new_path.exists() && new_path != *file_path;

            if collision {
                debug!("collision detected for: {}", new_name);
            }

            operations.push(RenameOperation::new(file_path.clone(), new_name, collision));
        }

        info!("prepared {} rename operations", operations.len());
        Ok(operations)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    use tempfile::TempDir;

    #[test]
    fn test_find_files_case_insensitive() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        // Create test files
        File::create(base_path.join("Resident.Alien.S01E01.srt")).unwrap();
        File::create(base_path.join("resident.alien.s01e02.srt")).unwrap();
        File::create(base_path.join("Other.Show.S01E01.srt")).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = service.find_files("resident").unwrap();

        assert_eq!(files.len(), 2);
    }

    #[test]
    fn test_find_files_only_srt() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        File::create(base_path.join("test.srt")).unwrap();
        File::create(base_path.join("test.txt")).unwrap();
        File::create(base_path.join("test.mp4")).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = service.find_files("test").unwrap();

        assert_eq!(files.len(), 1);
        assert!(files[0].to_str().unwrap().ends_with("test.srt"));
    }

    #[test]
    fn test_find_files_sorted() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        File::create(base_path.join("c.srt")).unwrap();
        File::create(base_path.join("a.srt")).unwrap();
        File::create(base_path.join("b.srt")).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = service.find_files(".srt").unwrap();

        assert_eq!(files.len(), 3);
        assert!(files[0].file_name().unwrap().to_str().unwrap() == "a.srt");
        assert!(files[1].file_name().unwrap().to_str().unwrap() == "b.srt");
        assert!(files[2].file_name().unwrap().to_str().unwrap() == "c.srt");
    }

    #[test]
    fn test_prepare_rename_operations_series_mode() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        let file1 = base_path.join("old1.srt");
        let file2 = base_path.join("old2.srt");
        File::create(&file1).unwrap();
        File::create(&file2).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = vec![file1.clone(), file2.clone()];

        let context = TemplateContext::new()
            .with_name(Some("Test Show".to_string()))
            .with_season(Some(1))
            .with_separator(Some(".".to_string()));

        let template = "{{ name }}.S{{ season }}E{{ episode }}.srt";
        let operations = service
            .prepare_rename_operations(files, template, &context, true)
            .unwrap();

        assert_eq!(operations.len(), 2);
        assert_eq!(operations[0].new_name, "Test Show.S01E01.srt");
        assert_eq!(operations[1].new_name, "Test Show.S01E02.srt");
    }

    #[test]
    fn test_prepare_rename_operations_collision_detection() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        let file1 = base_path.join("old.srt");
        let existing = base_path.join("new.srt");
        File::create(&file1).unwrap();
        File::create(&existing).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = vec![file1.clone()];

        let context = TemplateContext::new();
        let template = "new.srt";

        let operations = service
            .prepare_rename_operations(files, template, &context, false)
            .unwrap();

        assert_eq!(operations.len(), 1);
        assert!(operations[0].collision);
    }

    #[test]
    fn test_template_with_optional_variables() {
        let temp_dir = TempDir::new().unwrap();
        let base_path = temp_dir.path();

        let file1 = base_path.join("old.srt");
        File::create(&file1).unwrap();

        let service = FileRenameService::new(base_path.to_path_buf());
        let files = vec![file1.clone()];

        let context = TemplateContext::new()
            .with_name(Some("Show".to_string()))
            .with_separator(Some(".".to_string()));

        let template = "{{ name }}{{ separator }}subtitle.srt";
        let operations = service
            .prepare_rename_operations(files, template, &context, false)
            .unwrap();

        assert_eq!(operations.len(), 1);
        assert_eq!(operations[0].new_name, "Show.subtitle.srt");
    }
}

File: pkg/rename/service.go
package rename

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"text/template"
)

// FileRenameService is the filesystem-backed implementation of Service. Was
// Rust struct FileRenameService.
type FileRenameService struct {
	baseDir string
}

// NewFileRenameService creates a FileRenameService with the given base
// directory. Was FileRenameService::new.
func NewFileRenameService(baseDir string) *FileRenameService {
	return &FileRenameService{baseDir: baseDir}
}

// FindFiles locates .srt files in the base directory whose names contain the
// mask (case-insensitive), returning them sorted alphabetically. Was
// FileRenameService::find_files.
func (s *FileRenameService) FindFiles(mask string) ([]string, error) {
	slog.Info(fmt.Sprintf("searching for files with mask: %s", mask))

	maskLower := strings.ToLower(mask)
	foundFiles := make([]string, 0)

	entries, err := os.ReadDir(s.baseDir)
	if err != nil {
		return nil, NewIoError(fmt.Errorf("failed to read directory: %w", err))
	}

	for _, entry := range entries {
		path := filepath.Join(s.baseDir, entry.Name())

		info, statErr := os.Stat(path)
		if statErr != nil || !info.Mode().IsRegular() {
			continue
		}

		ext := filepath.Ext(path)
		if ext == "" {
			continue
		}
		// Rust compares the extension (without the leading dot) to exactly
		// "srt" (case-sensitive).
		if strings.TrimPrefix(ext, ".") != "srt" {
			continue
		}

		filename := filepath.Base(path)
		if strings.Contains(strings.ToLower(filename), maskLower) {
			slog.Debug(fmt.Sprintf("found file: %q", path))
			foundFiles = append(foundFiles, path)
		}
	}

	if len(foundFiles) == 0 {
		return nil, NewNoFilesFound(mask)
	}

	// Sort alphabetically (matches Rust PathBuf sort over full paths).
	sort.Strings(foundFiles)

	slog.Info(fmt.Sprintf("found %d files", len(foundFiles)))
	return foundFiles, nil
}

// PrepareRenameOperations renders the new file name for each input file using
// the template and context, detecting collisions. When seriesMode is set, an
// auto-incrementing 1-based `episode` variable is injected. Was
// FileRenameService::prepare_rename_operations.
func (s *FileRenameService) PrepareRenameOperations(
	files []string,
	tmplStr string,
	context *TemplateContext,
	seriesMode bool,
) ([]RenameOperation, error) {
	slog.Info("preparing rename operations")
	slog.Debug(fmt.Sprintf("series_mode: %t", seriesMode))
	slog.Debug(fmt.Sprintf("template: %s", tmplStr))

	// Replace tera with text/template. missingkey=error makes referencing an
	// absent variable a render error, mirroring tera's behavior.
	tmpl, err := template.New("filename").Option("missingkey=error").Parse(tmplStr)
	if err != nil {
		return nil, NewInvalidTemplate(err.Error())
	}

	operations := make([]RenameOperation, 0, len(files))

	for index, filePath := range files {
		ctx := make(map[string]any)

		if context.Name != nil {
			ctx["name"] = *context.Name
		}

		if context.Season != nil {
			ctx["season"] = fmt.Sprintf("%02d", *context.Season)
		}

		if context.Language != nil {
			ctx["language"] = *context.Language
		}

		if context.Separator != nil {
			ctx["separator"] = *context.Separator
		}

		if seriesMode {
			episodeNumber := index + 1
			ctx["episode"] = fmt.Sprintf("%02d", episodeNumber)
			slog.Debug(fmt.Sprintf("series mode: file %d -> episode %d", index, episodeNumber))
		}

		var sb strings.Builder
		if execErr := tmpl.Execute(&sb, ctx); execErr != nil {
			return nil, NewTemplateError(execErr.Error())
		}
		newName := sb.String()

		slog.Debug(fmt.Sprintf("new name for %q: %s", filePath, newName))

		newPath := filepath.Join(s.baseDir, newName)
		_, statErr := os.Stat(newPath)
		exists := statErr == nil
		collision := exists && newPath != filePath

		if collision {
			slog.Debug(fmt.Sprintf("collision detected for: %s", newName))
		}

		operations = append(operations, NewRenameOperation(filePath, newName, collision))
	}

	slog.Info(fmt.Sprintf("prepared %d rename operations", len(operations)))
	return operations, nil
}
